examples/simple.rs

```diff
@@ -1,46 +1,173 @@
 extern crate wayland_server;
 extern crate smithay;
+#[macro_use]
 extern crate glium;
 
+#[macro_use]
+extern crate slog;
+extern crate slog_async;
+extern crate slog_term;
+
+mod helpers;
+
 use glium::Surface;
+
+use helpers::{GliumDrawer, WlShellStubHandler};
+use slog::*;
+
 use smithay::backend::graphics::glium::IntoGlium;
 use smithay::backend::input::InputBackend;
 use smithay::backend::winit;
-use smithay::shm::ShmGlobal;
-use wayland_server::protocol::wl_shm;
+use smithay::compositor::{self, CompositorHandler, CompositorToken, TraversalAction};
+use smithay::shm::{BufferData, ShmGlobal, ShmToken};
+use wayland_server::{Client, EventLoopHandle, Liveness, Resource};
+
+use wayland_server::protocol::{wl_compositor, wl_shell, wl_shm, wl_subcompositor, wl_surface};
+
+struct SurfaceHandler {
+    shm_token: ShmToken,
+}
+
+#[derive(Default)]
+struct SurfaceData {
+    buffer: Option<(Vec<u8>, (u32, u32))>,
+}
+
+impl compositor::Handler<SurfaceData> for SurfaceHandler {
+    fn commit(&mut self, evlh: &mut EventLoopHandle, client: &Client, surface: &wl_surface::WlSurface,
+              token: CompositorToken<SurfaceData, SurfaceHandler>) {
+        // we retrieve the contents of the associated buffer and copy it
+        token.with_surface_data(surface, |attributes| {
+            match attributes.buffer.take() {
+                Some(Some((buffer, (x, y)))) => {
+                    self.shm_token.with_buffer_contents(&buffer, |slice, data| {
+                        let offset = data.offset as usize;
+                        let stride = data.stride as usize;
+                        let width = data.width as usize;
+                        let height = data.height as usize;
+                        let mut new_vec = Vec::with_capacity(width * height * 4);
```

Comment:
Does this happen whenever the buffer contents to change?
We should try to optimize this in the future for sure.